File: src/shared/types/attachment.ts
/**
 * Attachment domain types.
 * Represents files and images attached to messages.
 */

import type { ISODateString, TransferStatus, UUID } from "./common";

/** Supported file categories */
export type FileCategory = "image" | "video" | "document" | "archive" | "other";

/** MIME type mapping to category */
export const MIME_CATEGORY_MAP: Record<string, FileCategory> = {
  "application/msword": "document",
  "application/pdf": "document",
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
    "document",
  "application/x-7z-compressed": "archive",
  "application/x-rar-compressed": "archive",
  "application/zip": "archive",
  "image/gif": "image",
  "image/jpeg": "image",
  "image/png": "image",
  "image/webp": "image",
  "text/csv": "document",
  "text/plain": "document",
  "video/mp4": "video",
  "video/webm": "video",
} as const;

/** Attachment metadata stored in database */
export interface Attachment {
  readonly checksum: string;
  readonly createdAt: ISODateString;
  readonly height: number | null;
  readonly id: UUID;
  readonly isImage: boolean;
  readonly messageId: UUID;
  readonly mimeType: string;
  readonly originalFilename: string;
  readonly sizeBytes: number;
  readonly storedFilename: string;
  readonly width: number | null;
}

/** Attachment with download state */
export interface AttachmentWithDownloadState extends Attachment {
  readonly downloadProgress: number;
  readonly downloadStatus: TransferStatus | null;
  readonly localPath: string | null;
}

/** Data for creating an attachment */
export interface CreateAttachmentInput {
  readonly filePath: string;
  readonly messageId: UUID;
  readonly originalFilename: string;
}

/** Upload progress information */
export interface UploadProgress {
  readonly attachmentId: UUID;
  readonly bytesUploaded: number;
  readonly percentage: number;
  readonly totalBytes: number;
}

/** Download entity */
export interface Download {
  readonly attachmentId: UUID;
  readonly completedAt: ISODateString | null;
  readonly createdAt: ISODateString;
  readonly id: UUID;
  readonly localPath: string | null;
  readonly progressBytes: number;
  readonly startedAt: ISODateString | null;
  readonly status: TransferStatus;
  readonly updatedAt: ISODateString;
}

/** Download progress event */
export interface DownloadProgress {
  readonly attachmentId: UUID;
  readonly bytesDownloaded: number;
  readonly bytesPerSecond: number;
  readonly downloadId: UUID;
  readonly estimatedTimeRemaining: number | null;
  readonly percentage: number;
  readonly totalBytes: number;
}

/** File validation result */
export interface FileValidationResult {
  readonly category: FileCategory;
  readonly error: string | null;
  readonly isValid: boolean;
  readonly sanitizedFilename: string;
}

/** Helper to get file category from MIME type */
export function getFileCategory(mimeType: string): FileCategory {
  return MIME_CATEGORY_MAP[mimeType] ?? "other";
}

/** Helper to check if file is an image */
export function isImageMimeType(mimeType: string): boolean {
  return mimeType.startsWith("image/");
}

/** Format file size for display */
export function formatFileSize(bytes: number): string {
  if (bytes === 0) {
    return "0 B";
  }
  const k = 1024;
  const sizes = ["B", "KB", "MB", "GB"];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return `${Number.parseFloat((bytes / k ** i).toFixed(1))} ${sizes[i]}`;
}

File: src-tauri/src/models/attachment.rs
//! Attachment domain model.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Transfer status for downloads/uploads.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TransferStatus {
    Pending,
    InProgress,
    Paused,
    Completed,
    Failed,
    Cancelled,
}

impl Default for TransferStatus {
    fn default() -> Self {
        Self::Pending
    }
}

/// Attachment metadata stored in database.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Attachment {
    pub id: Uuid,
    pub message_id: Uuid,
    pub original_filename: String,
    pub stored_filename: String,
    pub mime_type: String,
    pub size_bytes: i64,
    pub checksum: String,
    pub is_image: bool,
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub created_at: DateTime<Utc>,
}

impl Attachment {
    /// Check if this is an image attachment.
    pub fn is_image_type(&self) -> bool {
        self.mime_type.starts_with("image/")
    }

    /// Check if this is a video attachment.
    pub fn is_video_type(&self) -> bool {
        self.mime_type.starts_with("video/")
    }
}

/// Download entity.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Download {
    pub id: Uuid,
    pub attachment_id: Uuid,
    pub status: TransferStatus,
    pub progress_bytes: i64,
    pub local_path: Option<String>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Download {
    /// Create a new pending download.
    pub fn new(attachment_id: Uuid) -> Self {
        let now = Utc::now();
        Self {
            id: Uuid::new_v4(),
            attachment_id,
            status: TransferStatus::Pending,
            progress_bytes: 0,
            local_path: None,
            started_at: None,
            completed_at: None,
            created_at: now,
            updated_at: now,
        }
    }
}

/// Download progress information.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DownloadProgress {
    pub download_id: Uuid,
    pub attachment_id: Uuid,
    pub bytes_downloaded: i64,
    pub total_bytes: i64,
    pub percentage: f64,
    pub bytes_per_second: f64,
    pub estimated_time_remaining: Option<i64>,
}

/// Input for creating an attachment.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateAttachmentInput {
    pub message_id: Uuid,
    pub file_path: String,
    pub original_filename: String,
}

/// Input for uploading files.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadFilesInput {
    pub file_paths: Vec<String>,
    pub message_id: Option<Uuid>,
}
